File: backend/app/schemas/incidents.py
"""Schemas for Incident Intelligence API."""

from typing import List, Optional
from pydantic import BaseModel
from backend.app.schemas.common import PaginationMeta, ProvenanceEnum


class IncidentRecord(BaseModel):
    """Incident intelligence record."""
    timestamp: str
    segment_id: str
    source_node: Optional[str] = None
    target_node: Optional[str] = None
    
    # Incident classification
    incident_state: str
    matched_incident_id: Optional[str] = None
    incident_type: Optional[str] = None
    severity: int
    lanes_blocked: int
    incident_lifecycle: Optional[str] = None
    incident_confidence: float
    
    # Underlying traffic state
    speed_kmh: float
    flow_vph: float
    queue_length_veh: float
    congestion_state: str
    traffic_evidence: Optional[str] = None
    incident_reasoning: Optional[str] = None
    provenance: str = ProvenanceEnum.DERIVED.value


class IncidentListResponse(BaseModel):
    """List of incident intelligence records with metadata."""
    meta: PaginationMeta
    data: List[IncidentRecord]

File: frontend/types/incidents.ts
/**
 * Types for Incident Intelligence API (GET /api/incidents).
 */

import { ProvenanceType, PaginationMeta } from "./api";

export type IncidentState =
  | "INCIDENT_SUPPORTED"
  | "INCIDENT_UNSUPPORTED"
  | "NO_INCIDENT_EVIDENCE"
  | "ROADWORK_SUPPORTED"
  | string;

export interface IncidentRecord {
  timestamp: string;
  segment_id: string;
  source_node?: string | null;
  target_node?: string | null;

  // Incident Classification
  incident_state: IncidentState;
  matched_incident_id?: string | null;
  incident_type?: string | null;
  severity: number; // 0, 1, 2, 3
  lanes_blocked: number;
  incident_lifecycle?: string | null;
  incident_confidence: number;

  // Traffic state evidence
  speed_kmh: number;
  flow_vph: number;
  queue_length_veh: number;
  congestion_state: string;
  traffic_evidence?: string | null;
  incident_reasoning?: string | null;
  provenance: ProvenanceType | string;
}

export interface IncidentListResponse {
  meta: PaginationMeta;
  data: IncidentRecord[];
}

export interface IncidentQueryParams {
  limit?: number;
  offset?: number;
  incident_state?: string;
  severity?: number;
}
